scripts/mark_eats/loom.lua

```diff
@@ -0,0 +1,1175 @@
+local screen_dirty = true
+local GRID_FRAMERATE = 30
+local grid_dirty = true
+local grid_leds = {}
+local grid_w, grid_h = 16, 8
+
+local beat_clock
+
+local grid_device
+local midi_in_device
+local midi_out_device
+local midi_out_channel
+
+local notes = {}
+local triggers = {}
+
+local root_note = 36
+local scale_type = 1
+local custom_scale = false
+local scale_edit_id = 1
+local scale_notes = {}
+local scale_note_names = {}
+local SCALES_LEN = #MusicUtil.SCALES
+
+local active_notes = {}
+local down_marks = {}
+local down_keys = {}
+local trails = {}
+local remove_animations = {}
+local DOWN_ANI_LENGTH = 0.2
+local REMOVE_ANI_LENGTH = 0.4
+local TRAIL_ANI_LENGTH = 6.0
+
+local pages
+local tabs
+local playback_icon
+local add_remove_animations = {}
+local ADD_REMOVE_ANI_LENGTH = 0.2
+local notes_changed_timeout = 0
+local triggers_changed_timeout = 0
+local NOTES_TRIGGERS_TIMEOUT_LENGTH = 0.2
+local save_data = {version = 1, patterns = {}}
+local save_menu_items = {"Load", "Save", "Delete"}
+local save_slot_list
+local save_menu_list
+local last_edited_slot = 0
+local confirm_message
+local confirm_function
+
+
+local function copy_object(object)
+  if type(object) ~= 'table' then return object end
+  local result = {}
+  for k, v in pairs(object) do result[copy_object(k)] = copy_object(v) end
+  return result
+end
+
+local function update_save_slot_list()
+  local entries = {}
+  for i = 1, math.min(#save_data.patterns + 1, 999) do
+    local entry
+    if i <= #save_data.patterns then
+      entry = save_data.patterns[i].name
+    else
+      entry = "-"
+    end
+    if i == last_edited_slot then entry = entry .. "*" end
+    entries[i] = i .. ". " .. entry
+  end
+  save_slot_list.entries = entries
+end
+
+local function read_data()
+  local disk_data = tab.load(DATA_FILE_PATH)
+  if disk_data and disk_data.version then
+    if disk_data.version == 1 then
+      save_data = disk_data
+    else
+      print("Unrecognized data, version " .. disk_data.version)
+    end
+  end
+  update_save_slot_list()
+end
+
+local function write_data()
+  tab.save(save_data, DATA_FILE_PATH)
+end
+
+local function load_pattern(index)
+  if index > #save_data.patterns then return end
+  
+  local pattern = copy_object(save_data.patterns[index])
+  params:set("bpm", pattern.bpm)
+  params:set("step_length", pattern.step_length)
+  params:set("pattern_width", pattern.pattern_width)
+  params:set("pattern_height", pattern.pattern_height)
+  params:set("min_velocity", pattern.min_velocity)
+  params:set("max_velocity", pattern.max_velocity)
+  notes = pattern.notes
+  triggers = pattern.triggers
+  root_note = pattern.root_note
+  scale_type = pattern.scale_type
+  custom_scale = pattern.custom_scale
+  scale_notes = pattern.scale_notes
+  scale_note_names = MusicUtil.note_nums_to_names(scale_notes, true)
+  
+  last_edited_slot = index
+  update_save_slot_list()
+  grid_dirty = true
+end
+
+local function save_pattern(index)
+  local pattern = {}
```

Comment:
ok, so here's a good candidate for a table constructor if you do decide to go that way...